# Level 1: Data Exploration & Basic Review

## Project Objective

The objective of Level 1 is to understand the structure and basic characteristics of the train schedule dataset. This includes reviewing the dataset size, attributes, train information, station details, and the number of stops associated with each train.

In [2]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

In [3]:
df = pd.read_csv("../DATASET/Dataset1.csv")

In [4]:
df.head()

,SN,Train_No,Station_Code,1A,2A,3A,SL,Station_Name,Route_Number,Arrival_time,Departure_Time,Distance
0,1,107,SWV,100,100,100,100,SAWANTWADI R,1,00:00:00,10:25:00,0
1,2,107,THVM,260,228,196,164,THIVIM,1,11:06:00,11:08:00,32
2,3,107,KRMI,345,296,247,198,KARMALI,1,11:28:00,11:30:00,49
3,4,107,MAO,490,412,334,256,MADGOAN JN.,1,12:10:00,00:00:00,78
4,1,108,MAO,100,100,100,100,MADGOAN JN.,1,00:00:00,20:30:00,0


In [5]:
df.shape

(186074, 12)

The dataset contains 186,074 records and 12 attributes. Each record represents a station-level entry associated with a train schedule.

In [6]:
# Dataset Structure
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 186074 entries, 0 to 186073
Data columns (total 12 columns):
 #   Column          Non-Null Count   Dtype
---  ------          --------------   -----
 0   SN              186074 non-null  int64
 1   Train_No        186074 non-null  int64
 2   Station_Code    186074 non-null  str  
 3   1A              186074 non-null  int64
 4   2A              186074 non-null  int64
 5   3A              186074 non-null  int64
 6   SL              186074 non-null  int64
 7   Station_Name    186074 non-null  str  
 8   Route_Number    186074 non-null  int64
 9   Arrival_time    186074 non-null  str  
 10  Departure_Time  186074 non-null  str  
 11  Distance        186074 non-null  int64
dtypes: int64(8), str(4)
memory usage: 17.0 MB


In [7]:
df.columns

Index(['SN', 'Train_No', 'Station_Code', '1A', '2A', '3A', 'SL',
       'Station_Name', 'Route_Number', 'Arrival_time', 'Departure_Time',
       'Distance'],
      dtype='str')

| Column           | Meaning                                       |
| ---------------- | --------------------------------------------- |
| `SN`             | Serial number of the record                   |
| `Train_No`       | Train number                                  |
| `Station_Code`   | Code identifying the station                  |
| `1A`             | 1st AC-related value                          |
| `2A`             | 2nd AC-related value                          |
| `3A`             | 3rd AC-related value                          |
| `SL`             | Sleeper-related value                         |
| `Station_Name`   | Name of the station                           |
| `Route_Number`   | Route sequence/identifier provided in dataset |
| `Arrival_time`   | Train arrival time at the station             |
| `Departure_Time` | Train departure time from the station         |
| `Distance`       | Distance associated with the station record   |


In [8]:
total_records = df.shape[0]
total_attributes = df.shape[1]

print("Total Records:", total_records)
print("Total Attributes:", total_attributes)

Total Records: 186074
Total Attributes: 12


In [9]:
total_trains = df["Train_No"].nunique()

print("Total Unique Trains:", total_trains)

Total Unique Trains: 11113


In [10]:
df["Train_No"].unique()

array([  107,   108,   128, ..., 99907, 99908, 22439], shape=(11113,))

In [19]:
df.groupby("Train_No").size().head(10)

Train_No
107     4
108     4
128    22
290    14
401    12
421     5
422     5
477    14
502     9
504     3
dtype: int64

In [20]:
df[df["Train_No"] == 107]

,SN,Train_No,Station_Code,1A,2A,3A,SL,Station_Name,Route_Number,Arrival_time,Departure_Time,Distance
0,1,107,SWV,100,100,100,100,SAWANTWADI R,1,00:00:00,10:25:00,0
1,2,107,THVM,260,228,196,164,THIVIM,1,11:06:00,11:08:00,32
2,3,107,KRMI,345,296,247,198,KARMALI,1,11:28:00,11:30:00,49
3,4,107,MAO,490,412,334,256,MADGOAN JN.,1,12:10:00,00:00:00,78


In [21]:
stops_per_train = df.groupby("Train_No")["Station_Code"].count()

stops_per_train.head()

Train_No
107     4
108     4
128    22
290    14
401    12
Name: Station_Code, dtype: int64

In [23]:
#Maximum and Minimum Stops
max_stops = stops_per_train.max()
min_stops = stops_per_train.min()

print("Maximum stops:", max_stops)
print("Minimum stops:", min_stops)

Maximum stops: 118
Minimum stops: 2


In [24]:
max_stop_trains = stops_per_train[
    stops_per_train == max_stops
]

min_stop_trains = stops_per_train[
    stops_per_train == min_stops
]

print("Train(s) with maximum stops:")
print(max_stop_trains)

print("\nTrain(s) with minimum stops:")
print(min_stop_trains)

Train(s) with maximum stops:
Train_No
53041    118
Name: Station_Code, dtype: int64

Train(s) with minimum stops:
Train_No
557      2
3305     2
3306     2
5715     2
5716     2
        ..
99499    2
99500    2
99501    2
99502    2
99503    2
Name: Station_Code, Length: 1249, dtype: int64


##  Level 1 Findings

- The dataset contains 186,074 records and 12 attributes.
- The dataset contains 11,113 unique trains.
- Each train is represented through its station-wise schedule records.
- Train 53041 has the maximum number of scheduled station records, with 118 stops.
- The minimum number of scheduled station records is 2.
- A total of 1,249 trains have the minimum of 2 scheduled station records.
- The analysis provides a basic understanding of train coverage, station-level records, and the variation in the number of stops across trains.